pkg/resource/generator.go

```diff
@@ -86,8 +93,125 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
+		return err
+	}
+	if len(name) == 0 {
+		runCreate = true
+	} else if driver.StorageChanged(cr) {
+		exists, err := driver.StorageExists(cr)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			runCreate = true
+		}
+	} else {
+		exists, err := driver.StorageExists(cr)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			runCreate = true
+		}
+	}
+
+	if runCreate {
+		if err := driver.CreateStorage(cr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// syncSecrets checks to see if we have updated storage credentials from:
+// 1.) user provided credentials in the image-registry-private-configuration-user secret
+// and updates the image-registry-private-configuration secret which provides
+// those credentials to the registry pod
+func (g *Generator) syncSecrets(cr *regopapi.ImageRegistry, modified *bool) error {
+	client, err := clusterconfig.GetCoreClient()
+	if err != nil {
+		return err
+	}
+
+	// Get the existing image-registry-private-configuration secret
+	sec, err := client.Secrets(g.params.Deployment.Namespace).Get(regopapi.ImageRegistryPrivateConfiguration, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("unable to get secret %q: %v", fmt.Sprintf("%s/%s", g.params.Deployment.Namespace, regopapi.ImageRegistryPrivateConfiguration), err)
+	}
+
+	// Create a driver with the current configuration
+	driver, err := storage.NewDriver(cr.Name, cr.Namespace, &cr.Spec.Storage)
+	if err != nil {
+		return err
+	}
+
+	data, err := driver.SyncSecrets(sec)
+	if err != nil {
+		return err
+	}
+	if data != nil {
+		glog.Infof("Updating secret %q with updated credentials.", fmt.Sprintf("%s/%s", g.params.Deployment.Namespace, regopapi.ImageRegistryPrivateConfiguration))
+
+		// Update the image-registry-private-configuration secret
+		upSec, err := util.CreateOrUpdateSecret(regopapi.ImageRegistryPrivateConfiguration, g.params.Deployment.Namespace, data)
+		if err != nil {
+			return err
+		}
+
+		// Generate a checksum of the updated secret
+		upSecChecksum, err := strategy.Checksum(upSec)
+		if err != nil {
+			return nil
+		}
+
+		// Don't create this client unless we need it
+		appsclient, err := kappsset.NewForConfig(g.kubeconfig)
+		if err != nil {
+			return err
+		}
+
+		// Get the Image Registry deployment
+		deployment, err := appsclient.Deployments(g.params.Deployment.Namespace).Get("image-registry", metaapi.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// Make sure that the annotations map isn't nil!
+		if deployment.Spec.Template.Annotations == nil {
+			deployment.Spec.Template.Annotations = make(map[string]string)
+		}
+
+		// Update the deployment template with an annotation of the checksum of the secret
+		// to trigger the Image Registry to be redeployed
+		deployment.Spec.Template.Annotations[fmt.Sprintf("%s-checksum", upSec.ObjectMeta.Name)] = upSecChecksum
+		if _, err := appsclient.Deployments(g.params.Deployment.Namespace).Update(deployment); err != nil {
```

Comment:
why isn't this done as part of the deployment generator/template?  Having two writers updating the deployment resource is liable to be problematic.

Reply:
Updated in current push